internal/storage/clickhouse.go

```diff
@@ -305,170 +307,162 @@ func (c *ClickHouseConnector) GetBlocks(qf QueryFilter) (blocks []common.Block,
 	return blocks, nil
 }
 
-func (c *ClickHouseConnector) GetTransactions(qf QueryFilter) (txs []common.Transaction, err error) {
+func (c *ClickHouseConnector) GetTransactions(qf QueryFilter) (QueryResult, error) {
 	columns := "chain_id, hash, nonce, block_hash, block_number, block_timestamp, transaction_index, from_address, to_address, value, gas, gas_price, input, max_fee_per_gas, max_priority_fee_per_gas, transaction_type"
-	query := fmt.Sprintf("SELECT %s FROM %s.transactions FINAL WHERE is_deleted = 0%s",
-		columns, c.cfg.Database, getLimitClause(int(qf.Limit)))
+	return c.executeQuery("transactions", columns, qf, scanTransaction)
+}
+
+func (c *ClickHouseConnector) GetLogs(qf QueryFilter) (QueryResult, error) {
+	columns := "chain_id, block_number, block_hash, block_timestamp, transaction_hash, transaction_index, log_index, address, data, topic_0, topic_1, topic_2, topic_3"
+	return c.executeQuery("logs", columns, qf, scanLog)
+}
+
+func (c *ClickHouseConnector) executeQuery(table, columns string, qf QueryFilter, scanFunc func(driver.Rows) (interface{}, error)) (QueryResult, error) {
```

Comment:
You could use a generic type here to have type checking working not only in runtime

Reply:
done